packages/node-sdk/src/client.ts

```diff
@@ -650,6 +412,341 @@ export class BucketClient {
+      "event context must be an object",
+    );
+    ok(
+      event.evalRuleResults === undefined ||
+        Array.isArray(event.evalRuleResults),
+      "event rule results must be an array",
+    );
+    ok(
+      event.evalMissingFields === undefined ||
+        Array.isArray(event.evalMissingFields),
+      "event missing fields must be an array",
+    );
+
+    const contextKey = new URLSearchParams(
+      flattenJSON(event.evalContext || {}),
+    ).toString();
+
+    if (this._config.offline) {
+      return;
+    }
+
+    if (
+      !this._config.rateLimiter.isAllowed(
+        hashObject({
+          action: event.action,
+          key: event.key,
+          targetingVersion: event.targetingVersion,
+          evalResult: event.evalResult,
+          contextKey,
+        }),
+      )
+    ) {
+      return;
+    }
+
+    await this._config.batchBuffer.add({
+      type: "feature-flag-event",
+      action: event.action,
+      key: event.key,
+      targetingVersion: event.targetingVersion,
+      evalContext: event.evalContext,
+      evalResult: event.evalResult,
+      evalRuleResults: event.evalRuleResults,
+      evalMissingFields: event.evalMissingFields,
+    });
+  }
+
+  /**
+   * Updates the context in Bucket (if needed).
+   * This method should be used before requesting feature flags or binding a client.
+   *
+   * @param options
+   */
+  private async syncContext(options: { enableTracking: boolean } & Context) {
+    if (!options.enableTracking) {
+      this._config.logger?.debug(
+        "tracking disabled, not updating user/company",
+      );
+
+      return;
+    }
+
+    const promises: Promise<void>[] = [];
+    if (typeof options.company?.id !== "undefined") {
+      const { id: _, ...attributes } = options.company;
+      promises.push(
+        this.updateCompany(options.company.id, {
+          attributes,
+          meta: { active: false },
+        }),
+      );
+    }
+
+    if (typeof options.user?.id !== "undefined") {
+      const { id: _, ...attributes } = options.user;
+      promises.push(
+        this.updateUser(options.user.id, {
+          attributes,
+          meta: { active: false },
+        }),
+      );
+    }
+
+    if (promises.length > 0) {
+      await Promise.all(promises);
+    }
+  }
+
+  /**
+   * Gets the features cache.
+   *
+   * @returns The features cache.
+   **/
+  private getFeaturesCache() {
+    if (!this._config.featuresCache) {
+      this._config.featuresCache = cache<FeaturesAPIResponse>(
+        this._config.refetchInterval,
+        this._config.staleWarningInterval,
+        this._config.logger,
+        async () => {
+          const res = await this.get<FeaturesAPIResponse>("features");
+
+          if (!isObject(res) || !Array.isArray(res?.features)) {
+            return undefined;
+          }
+
+          return res;
+        },
+      );
+    }
+
+    return this._config.featuresCache;
+  }
+
```

Comment:
since this was moved, it's hard to know if anything actually changed :/

Reply:
Nothing changed there and sorry, my IDE wants to be smart and group things by visibility. Might want to tweak that so it's not doing that.